common/string_helpers.cpp

```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
           ret.push_back(16 * *c1 + *c2);
           break;
         }
-        case 'u':
-          CARBON_FATAL() << "\\u is not yet supported in string literals";
+        case 'u': {
+          i++;
+          if (i >= source.size() || source[i] != '{')
+            return std::nullopt;
+          unsigned int unicode_int = 0;
+          i++;
+          while (i < source.size() && source[i] != '}') {
+            char hex_val;
+            if (source[i] >= '0' && source[i] <= '9') {
+              hex_val = source[i] - '0';
+            } else if (source[i] >= 'A' && source[i] <= 'F') {
+              hex_val = source[i] - 'A' + 10;
+            } else {
+              return std::nullopt;
+            }
+            unicode_int = unicode_int << 4;
+            unicode_int += hex_val;
+            i++;
+          }
+          if (i >= source.size())
+            return std::nullopt;
+          char temp[4];
+          char *result_two = &temp[0];
+          char *result = &temp[0];
+          bool conversion_ok = llvm::ConvertCodePointToUTF8(unicode_int, result_two);
+          if (!conversion_ok) {
+            return std::nullopt;
+          }
+          while(result < result_two) {
+            ret.push_back(*result);
+            result++;
+          }
```

Comment:
```suggestion
          ret.append(temp, result_two - temp);
```

Note we can use pointer math and `append` to avoid writing a loop. With this, I think you can also eliminate the `result` pointer.

Reply:
This change threw an error for me while testing, so I will _temporarily_ revert this back. I'll check on why it didn't work because your suggestion may be valid, but it may need a few more modifications to get it to work.

Is it because you missed the `result` variable? That's what the error message showed, so I am assuming that to be the case. Please correct me if I am wrong on that.

Reply:
@jonmeow , thanks for the suggestion! I was originally trying to do something like this, but I guess I was rusty on pointers, because I couldn't quite implement it.

Reply:
I managed to get this working by eliminating `char *result = &temp[0];`, though I am not sure if that's what was supposed to be done.